src/transaction_builder.js

```diff
@@ -153,33 +172,252 @@ export class TransactionBuilder {
-      throw new Error('timeout cannot be negative');
+    if (t < 0) {
+      throw new Error('max_time cannot be negative');
     }
 
-    if (timeout > 0) {
-      const timeoutTimestamp = Math.floor(Date.now() / 1000) + timeout;
-      if (this.timebounds === null) {
-        this.timebounds = { minTime: 0, maxTime: timeoutTimestamp };
-      } else {
-        this.timebounds = {
-          minTime: this.timebounds.minTime,
-          maxTime: timeoutTimestamp
-        };
-      }
-    } else {
+    if (this.timebounds === null) {
       this.timebounds = {
         minTime: 0,
         maxTime: 0
       };
     }
 
+    this.timebounds.maxTime = t;
+
+    return this;
+  }
+
+  /**
+   * If you want to prepare a transaction which will be valid at or after some ledger number in the future, you can set a minimum ledger precondition.
+   * Internally this will set the <code>minLedger</code> precondition.
+   *
+   * @param {number} minLedger The minimum ledger this transaction is valid at or after. Cannot be negative.
+   * If the value is `0`, the transaction is valid immediately.
+   * @return {TransactionBuilder}
+   * @see TransactionPreconditions
+   */
+  setMinLedger(minLedger) {
+    if (this.ledgerbounds !== null && this.ledgerbounds.minLedger > 0) {
+      throw new Error(
+        'LedgerBounds.min_ledger has been already set - setting min_ledger would overwrite it.'
+      );
+    }
+
+    if (minLedger < 0) {
+      throw new Error('min_ledger cannot be negative');
+    }
+
+    if (this.ledgerbounds === null) {
+      this.ledgerbounds = {
+        minLedger: 0,
+        maxLedger: 0
+      };
+    }
+
+    this.ledgerbounds.minLedger = minLedger;
+
+    return this;
+  }
+
+  /**
+   * If you want to prepare a transaction which will be valid before some ledger number in the future, you can set a maximum ledger precondition.
+   * Internally this will set the <code>maxLedger</code> precondition.
+   *
+   * @param {number} maxLedger The maximum ledger this transaction is valid before. Cannot be negative.
+   * If the value is `0`, the transaction is valid indefinitely.
+   * @return {TransactionBuilder}
+   * @see TransactionPreconditions
+   */
+  setMaxLedger(maxLedger) {
+    if (this.ledgerbounds !== null && this.ledgerbounds.maxLedger > 0) {
+      throw new Error(
+        'LedgerBounds.max_ledger has been already set - setting max_ledger would overwrite it.'
+      );
+    }
+
+    if (maxLedger < 0) {
+      throw new Error('max_ledger cannot be negative');
+    }
+
+    if (this.ledgerbounds === null) {
+      this.ledgerbounds = {
+        minLedger: 0,
+        maxLedger: 0
+      };
+    }
+
+    this.ledgerbounds.maxLedger = maxLedger;
+
+    return this;
+  }
+
+  /**
+   * If you want to prepare a transaction which will be valid only while the account sequence number is `minAccountSequence <= sourceAccountSequence < tx.seqNum`. Note that after execution the account's sequence number is always raised to `tx.seqNum`.
```

Comment:
nit: can we wrap these all to 80 or 100 chars? :sweat_smile: surprised the linter didn't do it